go/vt/vtgate/plugin_mysql_server.go

```diff
@@ -230,26 +230,42 @@ func newMysqlUnixSocket(address string, authServer mysql.AuthServer, handler mys
 	}
 }
 
-func init() {
-	servenv.OnRun(initMySQLProtocol)
+func shutdownMysqlProtocolAndDrain() {
+	if mysqlListener != nil {
+		mysqlListener.Close()
+		mysqlListener = nil
+	}
+	if mysqlUnixListener != nil {
+		mysqlUnixListener.Close()
+		mysqlUnixListener = nil
+	}
 
-	servenv.OnTermSync(func() {
-		if mysqlListener != nil {
-			mysqlListener.Close()
-			mysqlListener = nil
-		}
-		if mysqlUnixListener != nil {
-			mysqlUnixListener.Close()
-			mysqlUnixListener = nil
+	if atomic.LoadInt32(&busyConnections) > 0 {
+		log.Infof("Waiting for all client connections to be idle (%d active)...", atomic.LoadInt32(&busyConnections))
+		start := time.Now()
+		reported := start
+		for time.Since(start) < 30*time.Second {
```

Comment:
There is a global timeout for `OnTermSync()` hooks controlled by `-onterm_timeout` (defaults to 10 seconds). Can we depend on that instead?